src/core/thread/network_data_notifier.cpp

```diff
@@ -152,6 +169,75 @@ void Notifier::HandleCoapResponse(Error aResult)
     }
 }
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+
+bool Notifier::IsEligibleForRouterRoleUpgrade(void) const
+{
+    bool    isEligible = false;
+    uint8_t activeRouterCount;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsChild() && Get<Mle::MleRouter>().IsRouterEligible());
+
+    // RouterUpgradeThreshold can be explicitly set to zero in some of
+    // cert tests to disallow device to become router.
+
+    VerifyOrExit(Get<Mle::MleRouter>().GetRouterUpgradeThreshold() != 0);
+
+    // Check the local network data to ensure that we are a border
+    // router providing IP connectivity and therefore eligible to
+    // request router role upgrade with `kBorderRouterRequst`
+    // status.
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
+
+    activeRouterCount = Get<RouterTable>().GetActiveRouterCount();
+    VerifyOrExit((activeRouterCount >= Get<Mle::MleRouter>().GetRouterUpgradeThreshold()) &&
+                 (activeRouterCount < Mle::kMaxRouters));
+
+    VerifyOrExit(Get<Leader>().CountBorderRouters(kRouterRoleOnly) < Mle::kRouterUpgradeBorderRouterRequestThreshold);
+    isEligible = true;
+
+exit:
+    return isEligible;
+}
+
+void Notifier::RequestRouterRoleIfEligible(void)
+{
+    VerifyOrExit(IsEligibleForRouterRoleUpgrade() && (mRouterRoleUpgradeTimeout == 0));
+
+    mRouterRoleUpgradeTimeout = Random::NonCrypto::GetUint8InRange(1, kRouterRoleUpgradeMaxTimeout + 1);
+    Get<TimeTicker>().RegisterReceiver(TimeTicker::kNetworkDataNotifier);
+
+    LogInfo("Requesting router role as BR after %u sec", mRouterRoleUpgradeTimeout);
+
+exit:
+    return;
+}
+
+void Notifier::HandleTimeTick(void)
+{
+    if (mRouterRoleUpgradeTimeout > 0)
+    {
+        mRouterRoleUpgradeTimeout--;
+    }
+
+    if (mRouterRoleUpgradeTimeout == 0)
```

Comment:
```suggestion
    if (mRouterRoleUpgradeTimeout > 0 && --mRouterRoleUpgradeTimeout == 0)
```
nit: a commonly used patter for HandleTimeTick.

Reply:
I prefer the existing one, since if we enter this method with `mRouterRoleUpgradeTimeout` being 0 already, we still want to go through the second `if`.

This should not happen with current use, but I feel it is better to behave correctly in this situation (in case code is changed in future which allows this pattern).

Reply:
> if we enter this method with `mRouterRoleUpgradeTimeout` being 0 already, we still want to go through the second `if`.

I think this is the very reason why `(mRouterRoleUpgradeTimeout > 0 && --mRouterRoleUpgradeTimeout == 0)` is a better pattern. 

`mRouterRoleUpgradeTimeout = 0` should actually mean that the timer is not set. So, we should not execute the timeout logic in such case. 

Thoughts?

Reply:
I think either way is fine. I changed it in the new push.